src/Dhall/Core.hs

```diff
@@ -1130,6 +1141,22 @@ normalize e = case e of
             App NaturalIsZero (NaturalLit n) -> BoolLit (n == 0)
             App NaturalEven (NaturalLit n) -> BoolLit (even n)
             App NaturalOdd (NaturalLit n) -> BoolLit (odd n)
+            App (App OptionalBuild t) k
+                | check     -> OptionalLit t (Data.Vector.fromList k')
```

Comment:
Minor suggestion: instead of using `Data.Vector.fromList` here, you can have `go` build a `Vector` directly:

```haskell
go (App (Var "Just") e') = pure e'
go (Var "Nothing")       = empty
```